File: HashTable/subset_or_not.cpp
#include <iostream>
#include <unordered_set>

bool is_subset(int arr1[], int arr2[], int s1, int s2){
  std::unordered_set<int> hash_set;
  for(int i = 0; i < s1; i++){
    hash_set.insert(arr1[i]);
  }

  for(int i = 0; i < s2; i++){
    if(hash_set.find(arr2[i]) == hash_set.end()){
      return false;
    }
  }
  return true;
}

int main(){
  int arr1[] = {11, 1, 13, 21, 3, 7};
  int arr2[] = {11, 3, 7, 1};
  
  int s1= sizeof(arr1)/sizeof(arr1[0]);
  int s2= sizeof(arr2)/sizeof(arr2[0]);

  bool res = is_subset(arr1, arr2, s1, s2);
  if(res){
    printf("arr2[] is a subset of arr1[]\n");
  } else {
    printf("arr2[] is not a subset of arr1[]\n");

  }
  return 0;
}

File: HashTable/subset_or_not.go
package main

import "fmt"

type HashSet map[int]struct{}

func NewHashSet() *HashSet {
	set := make(HashSet)
	return &set
}

func (set *HashSet) Add(value int) {
	(*set)[value] = struct{}{}
}

func (set *HashSet) Remove(value int) {
	delete(*set, value)
}

func (set *HashSet) Contains(value int) bool {
	_, found := (*set)[value]
	return found
}

func isSubset(set *HashSet, arr []int) bool {
	for _, val := range arr {
		if !set.Contains(val) {
			return false
		}
	}
	return true
}

func main() {
	arr1 := []int{11, 1, 13, 21, 3, 7}
	arr2 := []int{11, 3, 7, 1}

	set1 := NewHashSet()

	for _, i := range arr1 {
		set1.Add(i)
	}

	res := isSubset(set1, arr2)

	if res == true {
		fmt.Println("yes it's subset")
	} else {
		fmt.Println("No it's not a subset")
	}
}
